## MNIST Classification 

- <font color="cyan">there are some methods which are used in multi-class classification for binary classifiers like SVM, Logistic Regression etc. You can read more about them here:</font> https://towardsdatascience.com/multi-class-classification-one-vs-all-one-vs-one-94daed32a87b   

#### Grid Search for Best Hyperparameters

In [1]:
from sklearn.datasets import load_digits
from sklearn.model_selection import GridSearchCV
from sklearn import svm

In [12]:
import multiprocessing 

In [13]:
cpus = multiprocessing.cpu_count()
print(cpus)

3


# Dataset Prepare

In [4]:
data = load_digits()
X = data.data
Y  = data.target
print(X.shape,Y.shape)

(1797, 64) (1797,)


## Logistic Regression
- Let us first try accuracy of logistic regression on MNIST dataset. Then we will try on SVM and compare them both.

In [6]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score

In [7]:
lr = LogisticRegression()
lr.fit(X,Y)
cross_val_score(lr,X,Y,scoring="accuracy",cv=5).mean()

/home/harsh/Machine_learning/ML_virtual_env/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/home/harsh/Machine_learning/ML_virtual_env/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stab

0.9137650882079852

### Logistic Regression gives around 92% accuracy

## SVM:
- Now let us try it on SVM

In [8]:
from sklearn import svm

In [9]:
svc = svm.SVC()
svc.fit(X,Y)
cross_val_score(svc,X,Y,scoring="accuracy",cv=5).mean()

0.9632838130609718

### Pretty Bad Result :(?
- This is bcoz svm uses default values of hyperparameters 'kernel' and 'C'. We can find **best hyperparameter using gridsearch method**

In [22]:
params = [
    {
        'kernel':['linear','rbf','poly','sigmoid'],
        'C':[0.1,0.2,0.5,1.0,2.0,5.0]
    }   
]

In [23]:
gs = GridSearchCV(estimator=svm.SVC(),param_grid=params,scoring="accuracy",cv=5,n_jobs = cpus)

In [24]:
gs.fit(X,Y)

GridSearchCV(cv=5, estimator=SVC(), n_jobs=3,
             param_grid=[{'C': [0.1, 0.2, 0.5, 1.0, 2.0, 5.0],
                          'kernel': ['linear', 'rbf', 'poly', 'sigmoid']}],
             scoring='accuracy')

In [25]:
gs.best_params_

{'C': 5.0, 'kernel': 'rbf'}

In [26]:
gs.best_score_

0.9738502011761063

## SVM with rbf kernel, C=5.0 gives 97% accuracy on the Dataset!